peer/peerlist/list.go

```diff
@@ -0,0 +1,553 @@
+	}
+}
+
+// New creates a new peer list with an identifier chooser for available peers.
+func New(name string, transport peer.Transport, availableChooser peer.RetainedChooserList, opts ...ListOption) *List {
+	cfg := defaultListConfig
+	for _, o := range opts {
+		o(&cfg)
+	}
+
+	return &List{
+		once:               lifecycle.NewOnce(),
+		name:               name,
+		uninitializedPeers: make(map[string]peer.Identifier, cfg.capacity),
+		unavailablePeers:   make(map[string]*peerThunk, cfg.capacity),
+		availablePeers:     make(map[string]*peerThunk, cfg.capacity),
+		availableChooser:   availableChooser,
+		transport:          transport,
+		peerAvailableEvent: make(chan struct{}, 1),
+	}
+}
+
+// List is a peer list which rotates which peers are to be selected in a circle
+type List struct {
+	lock sync.RWMutex
+
+	name string
+
+	shouldRetainPeers  atomic.Bool
+	uninitializedPeers map[string]peer.Identifier
+
+	unavailablePeers   map[string]*peerThunk
+	availablePeers     map[string]*peerThunk
+	availableChooser   peer.RetainedChooserList
+	peerAvailableEvent chan struct{}
+	transport          peer.Transport
+
+	once *lifecycle.Once
+}
+
+// Update applies the additions and removals of peer Identifiers to the list
+// it returns a multi-error result of every failure that happened without
+// circuit breaking due to failures.
+func (pl *List) Update(updates peer.ListUpdates) error {
+	if len(updates.Additions) == 0 && len(updates.Removals) == 0 {
+		return nil
+	}
+
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+
+	if pl.shouldRetainPeers.Load() {
+		return pl.updateInitialized(updates)
+	}
+	return pl.updateUninitialized(updates)
+}
+
+// updateInitialized applies peer list updates when the peer list
+// is able to retain peers, putting the updates into the available
+// or unavailable containers.
+//
+// Must be run inside a mutex.Lock()
+func (pl *List) updateInitialized(updates peer.ListUpdates) error {
+	var errs error
+	for _, peerID := range updates.Removals {
+		errs = multierr.Append(errs, pl.removePeerIdentifier(peerID))
+	}
+
+	for _, peerID := range updates.Additions {
+		errs = multierr.Append(errs, pl.addPeerIdentifier(peerID))
+	}
+	return errs
+}
+
+// updateUninitialized applies peer list updates when the peer list
+// is **not** able to retain peers, putting the updates into a single
+// uninitialized peer list.
+//
+// Must be run inside a mutex.Lock()
+func (pl *List) updateUninitialized(updates peer.ListUpdates) error {
+	var errs error
+	for _, peerID := range updates.Removals {
+		if _, ok := pl.uninitializedPeers[peerID.Identifier()]; ok {
+			delete(pl.uninitializedPeers, peerID.Identifier())
+		} else {
+			errs = multierr.Append(errs, peer.ErrPeerRemoveNotInList(peerID.Identifier()))
+		}
+	}
+	for _, peerID := range updates.Additions {
+		pl.uninitializedPeers[peerID.Identifier()] = peerID
+	}
+
+	return errs
+}
+
+// Must be run inside a mutex.Lock()
+func (pl *List) addPeerIdentifier(pid peer.Identifier) error {
```

Comment:
do we need to check that this pid isn't already in our list?

Reply:
addressed. thank you.